router/pkg/pubsub/kafka/kafka.go

```diff
@@ -0,0 +1,380 @@
+package kafka
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"github.com/twmb/franz-go/pkg/kerr"
+	"github.com/twmb/franz-go/pkg/kgo"
+	"github.com/wundergraph/cosmo/router/pkg/pubsub"
+	"github.com/wundergraph/graphql-go-tools/v2/pkg/engine/datasource/pubsub_datasource"
+	"github.com/wundergraph/graphql-go-tools/v2/pkg/engine/resolve"
+	"go.uber.org/zap"
+	"sync"
+	"time"
+)
+
+var (
+	_ pubsub_datasource.KafkaConnector = (*connector)(nil)
+	_ pubsub_datasource.KafkaPubSub    = (*kafkaPubSub)(nil)
+	_ pubsub.Lifecycle                 = (*kafkaPubSub)(nil)
+
+	errClientClosed = errors.New("client closed")
+)
+
+type record struct {
+	topic string
+	data  []byte
+}
+
+type subscription struct {
+	topics  []string
+	updater resolve.SubscriptionUpdater
+	context context.Context
+}
+
+type connector struct {
+	writeClient *kgo.Client
+	opts        []kgo.Opt
+	logger      *zap.Logger
+}
+
+func NewConnector(logger *zap.Logger, opts []kgo.Opt) (pubsub_datasource.KafkaConnector, error) {
+
+	writeClient, err := kgo.NewClient(append(opts,
+		// For observability, we set the client ID to "router"
+		kgo.ClientID("cosmo.router.producer"))...,
+	)
+	if err != nil {
+		return nil, fmt.Errorf("failed to create write client for Kafka: %w", err)
+	}
+
+	return &connector{
+		writeClient: writeClient,
+		opts:        opts,
+		logger:      logger,
+	}, nil
+}
+
+func (c *connector) New(ctx context.Context) pubsub_datasource.KafkaPubSub {
+	ps := &kafkaPubSub{
+		ctx:           ctx,
+		logger:        c.logger.With(zap.String("pubsub", "kafka")),
+		opts:          c.opts,
+		subscriptions: make(map[string][]*subscription),
+		clients:       make(map[string]*kgo.Client),
+		writeClient:   c.writeClient,
+		mu:            sync.RWMutex{},
+		closeWg:       sync.WaitGroup{},
+	}
+
+	return ps
+}
+
+// kafkaPubSub is a Kafka pubsub implementation.
+// It uses the franz-go Kafka client to consume and produce messages.
+// The pubsub is stateless and does not store any messages.
+// It uses a single write client to produce messages and a client per topic to consume messages.
+// Each client polls the Kafka topic for new records and updates the subscriptions with the new data.
+type kafkaPubSub struct {
+	ctx           context.Context
+	opts          []kgo.Opt
+	logger        *zap.Logger
+	clients       map[string]*kgo.Client
+	subscriptions map[string][]*subscription
+	writeClient   *kgo.Client
+	mu            sync.RWMutex
```

Comment:
It seems like you're using the mutext for both subscriptions and clients.
I guess at least for clients we should have it?
In general, I'd always use one mutex per object (map, slice) that's being protected by it.